python/aot/c_codegen.py

```diff
@@ -0,0 +1,208 @@
+import binascii
+from dataclasses import dataclass
+from typing import Sequence, Tuple
+
+from .abstract_values import TRITON_TO_C_TYPES
+from .sig_annotation_dsl import SignatureTokens, tokenize_signature_annotation
+
+THREADS_PER_WARP = 32
+
+
+def py_str_to_uchar_array(txt: str) -> Tuple[str, int]:  # (c_code, array len)
+    """Hexdump as string into a C array"""
+    hex_ = str(binascii.hexlify(bytes(txt, "utf-8")))[2:-1]
+    it = iter(hex_)
+    data = ", ".join([f"0x{x}{next(it)}" for x in it])
+    return data, len(hex_)
+
+
+common_header = """
+#pragma once
+
+#include <stdio.h>
+#include <stdint.h>
+#include <inttypes.h>
+#include <cuda.h>
+
+typedef struct
+{
+    int gX;
+    int gY;
+    int gZ;
+    int numWarps;
+} GridWarps;
+
+"""
+
+
+@dataclass
+class KernelSignatureData:
+    pointer_names: Sequence[str]
+    pointer_types: Sequence[str]
+    attribute_names: Sequence[str]
+    attribute_types: Sequence[str]
+    attr_size_vars: Sequence[str]
+
+    @property
+    def total_kernel_arguments(self):
+        return len(self.pointer_names) + len(self.attribute_names)
+
+    @property
+    def all_argument_names(self):
+        return self.pointer_names + self.attribute_names
+
+    @classmethod
+    def parse_sig_dsl(cls, sig_ann: str, arg_names: Sequence[str]):
+        tokens: SignatureTokens  = tokenize_signature_annotation(sig_ann, arg_names)
+        return KernelSignatureData(
+            pointer_names=tokens.pointer_names,
+            pointer_types=tokens.pointers,
+            attribute_names=tokens.attribute_names,
+            attribute_types=tokens.attributes,
+            attr_size_vars=tokens.attribute_sizes
+        )
+
+
+
+# def make_kernel_signature(poiniter_types: Sequence[str], attribute_types: Sequence[str], arg_names: Sequence[str]):
+def make_kernel_signature(sig_data: KernelSignatureData):
+    """
+    Assuming that kernel signature follow the following pattern (pointers..., attribure..., constexpr...)
+    """
+    pointers = [f"CUdeviceptr {arg}" for arg in sig_data.pointer_names]
+    attributes = []
+    for ty, arg in zip(sig_data.attribute_types, sig_data.attribute_names):
+        dtype = TRITON_TO_C_TYPES[ty]
+        attributes.append(f"{dtype} {arg}")
+
+    return ", ".join(pointers + attributes)
+
+
+# def _make_kernel_header_source(name: str, arg_names: Sequence[str], ptr_types: Sequence[str], attr_types: Sequence[str], hex_, bin_size: int, threads_per_warp: int) -> Tuple[str, str]:
+def _make_kernel_header_source(
+    name: str, sig_data: KernelSignatureData, hex_, bin_size: int, threads_per_warp: int
+) -> Tuple[str, str]:
+    kernel_signature = make_kernel_signature(sig_data)
+    arg_pointers = ", ".join(
+        [f"&{arg}" for arg in sig_data.pointer_names + sig_data.attribute_names]
+    )
+    num_args = sig_data.total_kernel_arguments
+
+    header = """
+unsigned char {name}_ptx[{bin_size}];
```

Comment:
I'd say we would even want to store cubin here rather than ptx -- since ptx will be just-compiled and this would defeat some of the purpose of an AoT compiler

Reply:
Totally agree, I'll make those adjustments once we will be happy with the general structure